src/Microsoft.VisualStudio.Composition/Reflection/SkipClrVisibilityChecks.cs

```diff
@@ -64,33 +65,64 @@ internal SkipClrVisibilityChecks(AssemblyBuilder assemblyBuilder, ModuleBuilder
         }
 
         /// <summary>
-        /// Ensures the CLR will skip visibility checks when accessing
-        /// the assembly that contains the specified member.
+        /// Gets the set of assemblies that a generated assembly must be granted the ability to skip visiblity checks for
+        /// in order to access the specified type.
         /// </summary>
-        /// <param name="memberInfo">The member that may not be publicly accessible.</param>
-        internal void SkipVisibilityChecksFor(MemberInfo memberInfo)
+        /// <param name="typeInfo">The type which may be internal.</param>
+        /// <returns>The set of names of assemblies to skip visibility checks for.</returns>
+        internal static ImmutableHashSet<AssemblyName> GetSkipVisibilityChecksRequirements(TypeInfo typeInfo)
         {
-            this.SkipVisibilityChecksFor(memberInfo.Module.Assembly);
+            Requires.NotNull(typeInfo, nameof(typeInfo));
+
+            if (ReflectionHelpers.IsPublic(typeInfo.AsType()))
+            {
+                return ImmutableHashSet<AssemblyName>.Empty;
+            }
+            else
+            {
+                var result = ImmutableHashSet<AssemblyName>.Empty
+                    .Add(typeInfo.Assembly.GetName());
+
+                // If this type has a base type defined in another assembly that is also internal
```

Comment:
Did you mean to leave the commented code here?

Reply:
Yes, since it really feels like it *should* be here, and I want it to remind me to ask the CLR folks why I don't need visibility into transitive assemblies when I only ask to skip visibility checks for the first assembly.

Reply:
Because what's happening there is that access checks are only done on the metadata as written. For instance, A derives from B derives from C. As written, A only knows about B, so the access check from A is only to see if it can access B. For access to C, the only access check is that B is allowed to access C. These rules are straightforward as long as no techniques are used to skirt around access checks, but when you do, you can see surprises like this.

Reply:
OK, thanks. If A defines a member that my proxy implements, evidently that doesn't require C's visibility into A. I guess if A defines a member with a Type that is internal to A, at that point C couldn't implement it without an explicit skip visibility check for A. Yes?